### PR TITLE
Initial implementation of friendlier errors

### PR DESCRIPTION
**Pull request by patrick91**
I'll add a proper description later, this PR is just for testing the CI steps.

Related: #2107

TODO:

- [x] assert that documentation exists (via pytest)
- [x] make all of this optional (flag and checking for imports)
- [x] Don't fail bad if ast parsing fails (for example with dynamic code)
- [x] don't add libcst as main dependency
- [x] print standard exception if we can't find the source (?)
- [x] redirect from errors.strawberry.rocks
- [x] add group to allow installing everything (strawberry-graphql[cli] should ok)

Future release:
- [ ] Errors in wasm (if libcst works)

### CODE REVIEW

**Comment by botberry** [comment]
# Pre-release
:wave:

Releasing commit [1c694722e2c8ac628ec8fb603029a1adf0c67aa2] to PyPi as pre-release! :package:

**Comment by botberry** [comment]
# Pre-release
:wave:

Pre-release **0.127.0.dev.1661177678** [1c694722e2c8ac628ec8fb603029a1adf0c67aa2] has been released on PyPi! :rocket:
You can try it by doing:

```shell
poetry add strawberry-graphql==0.127.0.dev.1661177678
```

File: strawberry/exceptions/invalid_argument_type.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+if TYPE_CHECKING:
+    from strawberry.types.fields.resolver import StrawberryResolver
+
+
+class InvalidArgumentTypeError(StrawberryException):
+    def __init__(
+        self,
+        resolver: "StrawberryResolver",
+        argument_name: str,
+        argument_type: Literal["union", "interface"],
```
Comment (BryceBeagle) [inline strawberry/exceptions/invalid_argument_type.py:19]: perhaps this should be an enum? Or maybe you should just pass the `StrawberryArgument` and hte exception can grab `argument.type`?

File: strawberry/exceptions/object_is_not_a_class.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        self.annotation_message = "function defined here"
+        self.suggestion = (
+            "To fix this error, make sure your use "
+            f"strawberry.{method_type.value} on a class."
+        )
+
+        super().__init__(self.message)
+
+    @classmethod
+    def input(cls, obj: object) -> "ObjectIsNotClassError":
```
Comment (BryceBeagle) [inline strawberry/exceptions/object_is_not_a_class.py:42]: Why not use `from __future__ import annotations`?

File: strawberry/exceptions/utils/getsource.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    if inspect.istraceback(object):
+        object = object.tb_frame
+    if inspect.isframe(object):
+        object = object.f_code
+    if inspect.iscode(object):
+        if not hasattr(object, "co_firstlineno"):
+            raise OSError("could not find function definition")
+        lnum = object.co_firstlineno - 1
+        pat = re.compile(
+            r"^(\s*def\s)|(\s*async\s+def\s)|(.*(?<!\w)lambda(:|\s))|^(\s*@)"
```
Comment (BryceBeagle) [inline strawberry/exceptions/utils/getsource.py:105]: oof

Comment (BryceBeagle) [reply strawberry/exceptions/utils/getsource.py:105]: also, perhaps this should be a `CONST`

File: tests/fields/test_resolvers.py [hunk 4]
```diff
@@ -102,105 +102,181 @@ def val(cls) -> str:
...
+        field_1: str = strawberry.field(name="field_1")
+        field_2: str = strawberry.field(name="field_2")
+        field_3: str = strawberry.field(name="field_3")
+        field_4: str = strawberry.field(name="field_4")
+        field_5: str = strawberry.field(name="field_5")
+        field_6: str = strawberry.field(name="field_6")
+        field_7: str = strawberry.field(name="field_7")
+        field_8: str = strawberry.field(name="field_8")
+        field_9: str = strawberry.field(name="field_9")
+        banana = strawberry.field(name="banana")
```
Comment (BryceBeagle) [inline tests/fields/test_resolvers.py:254]: 🍌

File: strawberry/exceptions/utils/getsource.py (hunk 3 above)
Comment (patrick91) [reply strawberry/exceptions/utils/getsource.py:105]: I'll check if we still use this code 😊

File: strawberry/exceptions/invalid_argument_type.py (hunk 1 above)
Comment (patrick91) [reply strawberry/exceptions/invalid_argument_type.py:19]: that's a good suggestion, I'll see if we can use StrawberryArgument, if not I'll go with enum 😊

File: strawberry/exceptions/utils/getsource.py (hunk 3 above)
Comment (patrick91) [reply strawberry/exceptions/utils/getsource.py]: it's gone :D

**Comment by botberry** [comment]
Thanks for adding the `RELEASE.md` file!



Here's a preview of the changelog:

---

This release...


---
Here's the preview release card for twitter:




Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to @patrick91 for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -143,10 +152,13 @@ jobs:
 
       - name: pytest
         run:
-          poetry run pytest --cov-report xml --cov=. --showlocals -vv -m
```
Comment (estyxx) [inline .github/workflows/test.yml:146]: just curios: what is the difference?

File: docs/errors.md [hunk 6]
```diff
@@ -0,0 +1,44 @@
+# Errors in strawberry
+
+Strawberry has a built-in errors for when something goes wrong with the creation
```
Comment (estyxx) [inline docs/errors.md:3]:
```suggestion
Strawberry has built-in errors for when something goes wrong with the creation
```

File: strawberry/exceptions/__init__.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
+from __future__ import annotations
```
[the remark this answers is not in the document]
Comment (estyxx) [reply strawberry/exceptions/__init__.py:1]: So those exceptions here are left over? 
Did plan to make it StrawberryException or we just don't care? 🤔

**Comment by patrick91** [comment]
from discord:

we should make this error better, if not already done here:

```
import strawberry
from strawberry.scalars import JSON
strawberry.union(
    name='NamedUnion',
    types=(
        # some other types
        JSON,
    ),
)
```

AttributeError: 'ScalarWrapper' object has no attribute '__name__'. Did you mean: '__ne__'.

**Pull request merged**